### ISSUE TITLE
Wrong decomposition?

### ISSUE THREAD

**Issue by jvoight** [assaferan/ModFrmGL2#5]
```
> AttachSpec("ModFrmGL2/ModFrmGL2.spec");
> H_N := sub<GL(2,Integers(25))|[[11,19,9,14],[4,16,7,22]]>;
> H := PSL2Subgroup(H_N); 
> M := ModularSymbols(H, 2, Rationals(), 0); 
> S := CuspidalSubspace(M); 
> D := Decomposition(S, HeckeBound(S));
> [Dimension(f)/2 : f in D];
[ 2, 2, 2, 2, 4, 4, 4, 8, 8 ]
> 
> T := HeckeOperator(S, 2);
> for p in PrimesUpTo(13) do
for>   if p eq 5 then continue; end if;
for>   T +:= (Random(2*p)-p)*HeckeOperator(S,p);
for>   Sort([<Degree(v[1]),v[2] div 2> : v in Factorization(CharacteristicPoly\
nomial(T))]);
for> end for;
[ <2, 3>, <2, 5>, <4, 1>, <8, 1>, <8, 1> ]
[ <1, 2>, <1, 2>, <2, 1>, <2, 1>, <2, 2>, <2, 2>, <4, 1>, <8, 1>, <8, 1> ]
[ <1, 2>, <1, 2>, <2, 1>, <2, 1>, <2, 2>, <2, 2>, <4, 1>, <8, 1>, <8, 1> ]
[ <1, 2>, <1, 2>, <1, 4>, <2, 1>, <2, 1>, <2, 2>, <4, 1>, <8, 1>, <8, 1> ]
[ <2, 1>, <2, 1>, <2, 1>, <2, 1>, <2, 2>, <2, 2>, <4, 1>, <8, 1>, <8, 1> ]
```

Seems to imply that the semisimple decomposition is
  (ab surf 1) x (ab surf 2) x (ab surf 3) x (ab surf 4) x (ab surf 5)^2 x (ab surf 6)^2 
  x (ab fourfold) x (ab 8-fold) x (ab 8-fold)
which means Decomposition is failing to mention that the fourfold is the square of an abelian surface?

**Comment by assaferan** [comment]
Decomposition cannot see the semisimple decomposition, as can be seen by 
> [assigned f`is_irreducible : f in D];
[ true, true, true, true, false, false, true, true, true ]

That is due to the presence of old forms, which Decomposition cannot detect, as it only decomposes the space as a module for the Hecke algebra, and these spaces are eigenspaces for all Hecke operators (at least away from the level).
The mysterious thing here is that the subgroup above seems to be maximal, so one does not expect to see old forms appearing here. It implies that we should expand our scope from subgroups of SL2(ZZ) to any subgroup which is commensurable with SL2(ZZ).

**Comment by jvoight** [comment]
I think this is a semisimple decomposition, or can be refined into one!  You just need to label the multiplicities.  In the extreme case, maybe all Hecke operators act by scalars on a 2-dimensional vector space.  I think this decomposes (arbitrarily) into two 1-dimensional eigenspaces, and that is a semisimple decomposition.  

The double spaces are weird here, but no one promised us multiplicity and a nice theory of oldforms!

**Comment by assaferan** [comment]
That's true, but that is not what the function Decomposition does. 
However, now that we have IsotypicDimensionDecomposition, we can run

> IsotypicDimensionDecomposition(S);
[ <2, 2>, <2, 2>, <2, 2>, <2, 2>, <2, 4>, <2, 4>, <4, 2>, <8, 2>, <8, 2> ]
[ true, true, true, true, false, false, true, true, true ]

and have a good guess.

**Comment by jvoight** [comment]
Yes, the question is to decide what Decomposition "does", or at least how to interpret the output.  This caused serious confusion.  

I propose that it output an additional boolean indicating if the eigenspaces are distinct, and some additional comments in the description of the function.  Or can we not edit these, because they're inherited from Magma and William Stein's code?

**Comment by assaferan** [comment]
They are inherited from Magma and William Stein's code, but maybe we should edit these. It already came up in a thread with Edgar that it might be wise to lose our attachments to the original ModSym package, and have this package as a standalone, making sure it does not override any Magma intrinsics.

**Comment by jvoight** [comment]
We're doing the same thing with our Belyi code right now--early on we assumed they would be stitched together, but seems really not to be the case.  So I agree, it's probably better to file for divorce.

**Comment by assaferan** [comment]
Now that we have field for divorce, it is possible to modify the Decomposition intrinsic.
However, I'm not sure anymore what is the expected output. If I have a subspace which is not irreducible (say corresponding to the abelian fourfold above), it decomposes into irreducible subspaces arbitrarily (in infinitely many ways). Should I just choose an arbitrary such decomposition?

**Comment by jvoight** [comment]
It's a good question!  

Another thing that is new is that we are now on the "inside", and we can perhaps convince the Magma team--which is at least partly composed of us--to do things in a better or different way.  :)  If we want to be backwards compatible, we could have this be a vararg, the isotypic decomposition is computed by default, and if the user wants a decomposition into irreducibles, they ask for it.  

If that makes sense, then think the answer to your question is yes?  Writing Hecke spaces as V = W^{oplus k} corresponds to writing the corresponding abelian variety as A ~ A_0^k up to isogeny over QQ which inherently is arbitrary if k > 1.  (The L-series of the result, however, remains intrinsic.)  Additional operators, like Atkin-Lehner involutions or level-lowering operators, may not preserve this decomposition--that's expected behavior, right?
